File: ch2/Merge_Sort_recursive/algorithm.cpp
#include <vector>
#include <iostream>

template<class T>
void merge(std::vector<T>& L, int low, int mid, int high) {
    typedef typename std::vector<T>::iterator iterator;

    iterator first = L.begin()+low;
    iterator between = L.begin()+mid;
    iterator second = L.begin()+mid+1;
    iterator last = L.begin()+high;

    std::vector<T> temp;

    while ((first<=between) && (second<=last)) {
        if (*first<*second) {
            temp.push_back(*(first++));         // because we want to increase the pointer not the value pointed ad
        } else {
            temp.push_back(*(second++));
        }
    }
    if (first<=between) { temp.insert(temp.end(),first,between+1); }    // the intervals are half open ... wasted alot of time on that
    else { temp.insert(temp.end(),second,last+1); }

    for(int i = 0;i<temp.size();i++) {      // I guess ... this could be much cleaner
        L[low+i]=temp[i];
    }
}

template<class T>
void split(std::vector<T>& L, int low, int high) {
    if (low < high) {
        int mid = (low+high)/2;
        split(L,low,mid);
        split(L,mid+1,high);
        merge(L,low,mid,high);
    }
}


int main() {
    std::vector<int> list = {5,3,9,10,8,2,1,5,32,1,-23,-4,1,23,-2,-49,-80,1,-3,-5,-123,-1,23,213,-999,12,-1293};
    split(list,0,list.size()-1);
    for (int i = 0 ; i<list.size();i++){
        std::cout<<list[i]<<"  ";
    }
    std::cout<<std::endl;
}

File: ch2/Merge_Sort_recursive/algorithm.go
package main

import "fmt"

func split(L []int, low int, high int){
	if low < high {
		mid:=(low+high)/2
		split(L, low, mid)
		split(L, mid+1, high)
		merge(L, low, mid, high)
	}
}

func merge(L []int, low int, mid int, high int) {
	first:= low
	second:= mid + 1
	temp:= []int{}
	for (first <= mid) && (second<=high) {
		if L[first] < L[second] {
			temp = append(temp, L[first])
			first = first + 1
		} else {
			temp = append(temp, L[second])
			second = second + 1
		}
	}
	if first <= mid {
		for first <= mid {
			temp = append(temp,L[first])
			first = first + 1
		}
	} else {
		for second <= high {
			temp = append(temp,L[second])
			second = second + 1
		}
	}
	for i:=low; i<=high; i++ {
		L[i] = temp[i]
	}
}

func main() {
	L := []int{5,4,6,7,1,3}
	split(L, 0, 5)
	fmt.Println(L)
}